src/AggregateFunctions/AggregateFunctionDeepMergeJSON.h

```diff
@@ -0,0 +1,114 @@
+#pragma once
+
+#include <map>
+#include <AggregateFunctions/AggregateFunctionFactory.h>
+#include <AggregateFunctions/IAggregateFunction.h>
+#include <Columns/ColumnObject.h>
+#include <DataTypes/DataTypeObject.h>
+#include <IO/ReadHelpers.h>
+#include <IO/WriteHelpers.h>
+#include <base/StringRef.h>
+#include <Common/Arena.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+}
+
+/// Reasonable limits to prevent memory abuse
+constexpr size_t MAX_JSON_MERGE_PATHS = 10000;
+constexpr size_t MAX_JSON_MERGE_PATH_LENGTH = 1000;
+constexpr size_t MAX_JSON_MERGE_TOTAL_SIZE = 100_MiB;
+
+/// JSON unset marker key
+constexpr const char * UNSET_KEY = "$unset";
+
+struct DeepMergeJSONAggregateData
+{
+    struct PathData
+    {
+        Field value;
+        size_t row_order;
+        // Track if this path was explicitly deleted
+        bool is_deleted = false;
```

Comment:
Why do we need to store deleted paths? If it was deleted, we need to remove it or its children from the data and that's it.